pkg/container/versioned/value.go

```diff
@@ -0,0 +1,337 @@
+}
+
+// AllHandle returns a handle that matches all current/non-removed values
+func AllHandle() Handle {
+	return newHandle(maxVersion)
+}
+
+func (h Handle) getVersion() Version {
+	return h.versionPlus1 - 1
+}
+
+func (h Handle) IsValid() bool {
+	return h.versionPlus1 != 0
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) Load() Version {
+	return Version(a.version.Load())
+}
+
+func (a *atomicVersion) Store(version Version) {
+	a.version.Store(uint64(version))
+}
+
+func (a *atomicVersion) CompareAndSwap(old, new Version) bool {
+	return a.version.CompareAndSwap(uint64(old), uint64(new))
+}
+
+// Manager manages the versioning for multiple VersionedValues.
+// It is to be stored wherever VersionedValues are stored
+type Manager struct {
+	version atomicVersion
+
+	// mutex protects 'handles'
+	mutex lock.RWMutex
+	// handles is an ordered list of outstanding handle versions. Outdated values can be removed
+	// when there are no outstanding handles for it.
+	handles []Version
+
+	// latest cleaned version
+	cleanedVersion Version
+
+	// cleaner is called with the earliest version that must be kept
+	// cleaner is called while 'mutex' is held!
+	cleaner func(keepVersion Version)
+}
+
+func NewVersionManager(cleaner func(keepVersion Version)) *Manager {
+	return &Manager{
+		cleaner: cleaner,
+	}
+}
+
+func (v *Manager) GetVersion() (current, next Version) {
+	current = v.version.Load()
+	return current, current + 1
+}
+
+func (v *Manager) PublishVersion(current, new Version) bool {
+	return v.version.CompareAndSwap(current, new)
+}
+
+type Closer func() bool
+
+// GetHandle returns a Handle for the current version and a closer for releasing old versions held
+// for this version.
+// 'name' is for debugging purposes only
+func (v *Manager) GetHandle(name string) (handle Handle, closer Closer) {
+	// version is monotonically increasing, so appending the current version to the end of the
+	// outstanding handles keeps it sorted
+	v.mutex.Lock()
+	version := v.version.Load()
+	handle = newHandle(version)
+	v.handles = append(v.handles, version)
+	v.mutex.Unlock()
+
+	return handle, func() bool {
+		if version == invalidVersion {
+			// 2nd call does nothing
+			return false
+		}
+
+		v.mutex.Lock()
+		n, found := slices.BinarySearch(v.handles, version)
+		if found {
+			v.handles = slices.Delete(v.handles, n, n+1)
+
+			// release memory for long empty slice
+			if len(v.handles) == 0 && cap(v.handles) > maxEmptyHandles {
+				v.handles = nil
+			}
+			version = invalidVersion
+		} else {
+			log.WithFields(logrus.Fields{
+				"version": handle.getVersion(),
+				"handle":  name,
+			}).Error("closer: Handle for version not found")
+
+			v.mutex.Unlock()
+			return false
+		}
+
+		// clean if needed
+		keepVersion := invalidVersion
+		if len(v.handles) > 0 {
+			keepVersion = v.handles[0]
+		}
+		// 'keepVersion' is the current version if there are no outstanding handles
+		if keepVersion == invalidVersion {
```

Comment:
isn't this just `else`? Or is having `v.handles[0]` be invalidVersion expected?

Reply:
not really expected, will simplify.